theano/sandbox/cuda/blas.py

```diff
@@ -2600,76 +2597,38 @@ def c_support_code_apply(self, node, nodename):
+                for (int r2 = 0; r2 < ds2 && (i2*st2 + r2 < xD2); ++r2)
                 {
-                    for(int col_iter = 0;
-                        (tx + col_iter * bdimx < xD3) ; col_iter++){
-
-                        //The if inside is to don't do the division if we
-                        // need only 1 col_iter
-
-                        if(tx + bdimx < xD3)
-                        {
-                            x_col = tx + col_iter * bdimx;
-                            z_col = x_col/ds1;
-                        }
-
-                        if (%(ignore_border)s && ((x_col >= ds1 * D3) || (i2 >= D2)))
-                        {
-                            // This happens only if x_col, or i2*ds0, was ignored
-                            // (via ignore_border)
-                            // TODO: if ignore_border is False, this is impossible
-                            //        and we don't even need to generate this code.
-
-                            my_gz = 0.0f;
-
-                            //any fp number suffices for my_z, so we don't even
-                            //need to set it to anything in particular.
-
-                        }
-                        else
-                        {
-                            // this is effectively:
-                            // my_gz = gz[image_row][image_col][z_row][z_col]
-                            // my_z  = z[image_row][image_col][z_row][z_col]
-                            my_gz = gz[i0 * gzS0 + i1 * gzS1 + i2 * gzS2 +
-                                       z_col*gzS3];
-                            my_z =   z[i0 *  zS0 + i1 *  zS1 + i2 *  zS2 +
-                                       z_col* zS3];
-                        }
-                        for (int x_row = i2*ds0;
-                              (x_row < i2*ds0+ds0) && (x_row < xD2); ++x_row)
+                    for (int r3 = 0; r3 < ds3 && (i3*st3 + r3 < xD3); ++r3)
+                    {
+                        x_row = i2*st2 + r2;
+                        x_col = i3*st3 + r3;
+                        if (my_z == x[i0*xS0 + i1*xS1 + x_row*xS2 + x_col*xS3])
                         {
-                            // this is effectively:
-                            // gx[image_row][image_col][x_row][x_col]
-                            //   = (my_z == x[image_row][image_col][
-                            //                x_row][x_col]) ? my_gz : 0.0f;
-                            gx[i0*gxS0 + i1*gxS1 + x_row*gxS2 + x_col*gxS3]
-                               = (my_z == x[i0*xS0 + i1*xS1 + x_row*xS2 +
-                                            x_col*xS3]) ? my_gz : 0.0f;
+                            my_gz = gz[i0*gzS0 + i1*gzS1 + i2*gzS2 + i3*gzS3];
+                            // for overlapping pooling regions concurrent
+                            // writes are possible
+                            atomicAdd(&gx[i0*gxS0 + i1*gxS1 + x_row*gxS2 + x_col*gxS3], my_gz);
```

Comment:
@abergeron do you think it is useful to make a condition and do an atomicAdd only when there is overlapping?